✨Update the Helm chart to support external cluster as an ITS CP

francostellari opened the pull request:
## Summary

Example usage:

```shell
kind create cluster --name ext1
clusteradm init

scripts/create-kind-cluster-with-SSL-passthrough.sh --name kubeflex --port 9443

scripts/create-external-bootstrap-secret.sh --controlplane its1 --source-context kind-ext1 --address https://ext1-control-plane:6443 --verbose

helm dependency update core-chart

helm upgrade --install ks-core core-chart \
  --set-json='ITSes=[{"name":"its1","type":"external","install_clusteradm":false}]' \
  --set-json='WDSes=[{"name":"wds1"}]'
```

Requirements: kubeflex >=0.8

## Related issue(s)

Fixes #2684

pdettori commented on core-chart/values.yaml:
```diff
@@ -75,14 +75,14 @@ transport_controller:
 ITSes: # all the CPs in this list will execute the its.yaml PCH
   # - name: its1 # name of the control plane
-  #   type: host # optional type of control plane host or vcluster (default to vcluster, if not specified)
+  #   type: host # optional type of control plane host, vcluster, or external (default to vcluster, if not specified)
```
s/plane host/plane: host/

pdettori commented on core-chart/values.yaml:
```diff
@@ -75,14 +75,14 @@ transport_controller:
   #   install_clusteradm: true|false  # optional flag to enable/disable the installation of OCM in the control plane (default to true, if not specified)
   # - name: its2 # name of the control plane
-  #   type: host # optional type of control plane host or vcluster (default to vcluster, if not specified)
+  #   type: host # optional type of control plane host, vcluster, or external (default to vcluster, if not specified)
```
s/plane host/plane: host/

pdettori commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +111,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
 ITSes: # all the CPs in this list will execute the its.yaml PCH
   - name: <its1>          # mandatory name of the control plane
-    type: <vcluster|host> # optional type of control plane host or vcluster (default to vcluster, if not specified)
+    type: <vcluster|host|external> # optional type of control plane host, vcluster, or external (default to vcluster, if not specified)
```
s/plane host/plane: host/

pdettori commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +111,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
+      key: <key-name> # default: "kubeconfig-incluster"
   - name: <its2>          # mandatory name of the control plane
-    type: <vcluster|host> # optional type of control plane host or vcluster (default to vcluster, if not specified)
+    type: <vcluster|host|external> # optional type of control plane host, vcluster, or external (default to vcluster, if not specified)
```
s/plane host/plane: host/

pdettori commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +111,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
-where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be either vcluster (default) or host, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control plane.
+where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be vcluster (default), host, or external, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control pl ...
+
+When the ITS `type` is `external`, the `bootstrapSecret` sub-section can be used to indicate the bootstrap secret used by KubeFlex to connect to the external cluster. Specifically, it can be sued to indicate any of the optional values for the name of the secret, the namespace containing the secret, and the name of the key containg the kubeconfig of the external cluster.
```
s/sued/used/

pdettori commented:
This looks great @francostellari ! it may be helpful to have the "example usage" somewhere in the docs.

pdettori commented:
while testing, I have run into this issue:

```console
$ scripts/create-external-bootstrap-secret.sh --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
bash: scripts/create-external-bootstrap-secret.sh: Permission denied
$ ls -al scripts/create-external-bootstrap-secret.sh
-rw-r--r--@ 1 paolo  staff  6206 Feb  3 10:48 scripts/create-external-bootstrap-secret.sh
```

Seems like you need to adjust permissions for the script.

pdettori commented:
I manually updated the script permissions to move forward, but then got into another hurdle:

```console
$ chmod u+x scripts/create-external-bootstrap-secret.sh
$ scripts/create-external-bootstrap-secret.sh --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
Extracting context kind-ext1 from kubeconfig /Users/paolo/.kube/config...
Setting server internal address to https://ext1-control-plane:6443...
Creating namespace default...
Error from server (AlreadyExists): namespaces "default" already exists
```

pdettori commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+        break
+    fi
+done
+if [[ $create_ns ]] ; then
```
The condition inside [[ $create_ns ]] checks if the create_ns variable is non-empty. In Bash, any non-empty string is considered true, so this is always executed.

francostellari replied:
I fixed it

pdettori commented:
After manual fix for the bootstrap secret script, I have run into this issue:

```console
$ helm dependency build core-chart
Error: the lock file (Chart.lock) is out of sync with the dependencies file (Chart.yaml). Please update the dependencies
```

is there any guidance to resolve this?

francostellari commented:
> ```
> helm upgrade --install ks-core core-chart \
>   --set-json='ITSes=[{"name":"its1","type":"external","install_clusteradm":false}]' \
>   --set-json='WDSes=[{"name":"wds1"}]'
> ```

I fixed the bug that I introduced when I made a change to the YAML.

francostellari commented:
> ```shell
> helm upgrade --install ks-core core-chart \
>   --set-json='ITSes=[{"name":"its1","type":"external","install_clusteradm":false}]' \
>   --set-json='WDSes=[{"name":"wds1"}]'
> ```

I have added a section to the doc to detail a usage example with kind and how to use the local chart.

pdettori commented on docs/content/direct/core-chart.md:
````diff
@@ -152,7 +152,14 @@ where `name` must specify a name unique among all the control planes in that Kub
+
+```shell
+helm dependency update core-chart
+helm upgrade --install core-chart
````
don't we need also `helm dependency build` ?

pdettori commented on docs/content/direct/core-chart.md:
````diff
@@ -193,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
     ```
 
+The core chart also supports the use of external clusters as ITS Control Planes. A specific application is to connect to existing OCM clusters. As an exmaple, create a first local kind cluster with OCM installed in it:
+
````
s/exmaple/example/

pdettori commented on docs/content/direct/core-chart.md:
````diff
@@ -193,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-external-bootstrap-secret.sh) --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
+```
+
+Note the above command creates a secret named `its1-bootstrap` will be created in the `default` namespace of the `kind-kubeflex` cluster. Finally, install the core chart using the `ext1` cluster as ITS:
````
s/creates a secret named `its1-bootstrap` will be created in/creates a secret named `its1-bootstrap` in/

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -24,9 +24,9 @@ to join Workload Execution Clusters (WECs) (_e.g._, clusteradm),
 For such purpose, a full list of executable that may be required can be found [here](./pre-reqs.md).
 
-The setup of KubeStellar via the Core chart requires the existence of a KubeFlex hosting cluster.
+The setup of KubeStellar via the Core chart requires the existence of a Kubernetes hosting cluster.
```
I think the original is better. We have a role called "KubeFlex hosting cluster", and that is what is needed here.

francostellari replied:
ok, changed it back

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -24,9 +24,9 @@ to join Workload Execution Clusters (WECs) (_e.g._, clusteradm),
+The setup of KubeStellar via the Core chart requires the existence of a Kubernetes hosting cluster.
 
-This can be:
+While not a complete list of supported hosting clusters, we systematically test our releases using:
```
Being more explicit about the fact that this list is not exhaustive is good.
However, all that we test in CI is `kind`. I am not sure that we need this sentence to try to make assertions about testing.

francostellari replied:
I changed the sentence trying to satisfy both your comment and Paolo's

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +111,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
-where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be either vcluster (default) or host, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control plane.
+where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be vcluster (default), host, or external, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control pl ...
+
+When the ITS `type` is `external`, the `bootstrapSecret` sub-section can be used to indicate the bootstrap secret used by KubeFlex to connect to the external cluster. Specifically, it can be used to indicate any of the optional values for the name of the secret, the namespace containing the secret, and the name of the key containg the kubeconfig of the external cluster.
```
The parsing of the second sentence here is not clear. At first I read it like this: "... used to indicate (a) any of the optional values for the name of the secret, (b) the namespace containing the secret, and (c) the name of the key containg the kubeconfig of the external cluster". I would rewrite it a little. Such as:

> Specifically, it can be used to indicate any of the optional values for: the name of the secret, the namespace containing the secret, and the name of the key containg the kubeconfig of the external cluster.

francostellari replied:
I made it absolute to the release version

francostellari replied:
It seems to me that the first way your read it was the correct way... perhaps I can add `any combination of the optional values...`  (a), (b), (c)

MikeSpreitzer replied:
The problem with my first reading is that it confines "any of the optional values" to (a) but I think you meant "any of the optional values" to apply to the list of (a), (b), and (c).

francostellari replied:
I changed the text

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -140,7 +152,14 @@ where `name` must specify a name unique among all the control planes in that Kub
+helm upgrade --install core-chart
+```
+
+Alternatively, a specific version of the KubeStellar core chart can be simply installed in an existing cluster using the following command:
````
The introduction a second way to instantiate the chart breaks the subject of the sentence on new line 168:

> The above command will ...

francostellari replied:
I changed it to `Either if of the previous way of installing KubeStellar chart will install`

MikeSpreitzer commented:
A preview of the updated web site would be nice.

MikeSpreitzer commented on core-chart/values.yaml:
```diff
@@ -75,14 +75,14 @@ transport_controller:
 ITSes: # all the CPs in this list will execute the its.yaml PCH
   # - name: its1 # name of the control plane
-  #   type: host # optional type of control plane host or vcluster (default to vcluster, if not specified)
+  #   type: host # optional type of control plane: host, vcluster, or external (default to vcluster, if not specified)
```
Shouldn't `bootstrapSecret` be described here too?

francostellari replied:
I added some comments here

MikeSpreitzer replied:
You do not seem to have pushed those changes yet.

francostellari replied:
it should be in the latest version

MikeSpreitzer replied:
I still do not see it. Here is what I see (showing commit ID):
<img width="1290" alt="image" src="https://github.com/user-attachments/assets/3cb25307-f0bb-4032-ab2e-da2d8164e8e9" />

francostellari replied:
Sorry, I misread your initial comment and kept thinking to the `core-chart.md`. I updated the `value.yaml` file this time

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -181,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
       --set-json='WDSes=[{name":"wds2"}]'
     ```
 
+The core chart also supports the use of external clusters as ITS Control Planes. A specific application is to connect to existing OCM clusters. As an example, create a first local kind cluster with OCM installed in it:
````
To my taste, this material should come before the material about "After the initial installation is completed, there are two main ways to install additional control planes", because this material is germane to the initial installation.

francostellari replied:
Good idea, I switched the chunks

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -181,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
       --set-json='WDSes=[{name":"wds2"}]'
     ```
 
+The core chart also supports the use of external clusters as ITS Control Planes. A specific application is to connect to existing OCM clusters. As an example, create a first local kind cluster with OCM installed in it:
````
At the KubeStellar layer of abstraction we speak of spaces rather than (KF) ControlPlanes. I would rewrite the initial the initial sentence to be something like the following (also switching from plurals to singular for clarity).

> The core chart also supports the use of a pre-existing cluster (or any space, really) as an ITS.

francostellari replied:
ok, changed

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -181,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-external-bootstrap-secret.sh) --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
+```
+
+Note the above command creates a secret named `its1-bootstrap` in the `default` namespace of the `kind-kubeflex` cluster. Finally, install the core chart using the `ext1` cluster as ITS:
````
There are two commands, so I would say something like "Note that the last command above creates ...".

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -181,6 +200,32 @@ After the initial installation is completed, there are two main ways to install
+```shell
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-kind-cluster-with-SSL-passthrough.sh) --name kubeflex --port 9443
+
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-external-bootstrap-secret.sh) --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
````
That `--address https://ext1-control-plane:6443` part is black magic unless explained. I might add an explanation along the following lines.

The `--address` URL needs to be one that the KubeFlex controller can use to open a connection to the external cluster's Kubernetes apiserver(s). In this example, the external cluster is a `kind` cluster with one kube-apiserver and it listens on port 6443. We rely on the DNS resolver in Docker networking to map the domain name `ext1-control-plane` to the address of the container of that same name on the relevant Docker network.

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+
+# Command line arguments
+arg_cp=""
+arg_kubeconfig="$HOME/.kube/config"
```
Shouldn't this be the following?
```suggestion
arg_kubeconfig="${KUBECONFIG:-$HOME/.kube/config}"
```

francostellari replied:
yes, thank you

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+###############################################################################
+if [[ -z "$arg_cp" ]] ; then
+    echoerr "a control plane name is required!"
+    $0 --help
```
In curl-to-bash style usage, `$0` will be `bash` rather than a pathname of this script, right?

francostellari replied:
Perhaps but we have used this way in all other scripts, I believe.
I'm not sure if there is a solution besides hardcoding it.
I think it's a minor gripe given the small amount of `cur|bash --help`.
People will care about the list of commands.
But I'm open to a better solution if available.

MikeSpreitzer replied:
Because we document using this script in curl-to-bash style, I think it should function properly when used in that style.

Since this script's handling of `--help` is simply to invoke `display_help`, that could be in-lined here.

francostellari replied:
ok, I misunderstood the first comment, I changed it now in all places

MikeSpreitzer replied:
Yes, I see that you changed it in some places. However, there are still uses of `$0` to refer to the script name. Go ahead and do a text search for "$0". It appears in error messages, which should not pin the blame on bash.

MikeSpreitzer replied:
I think that we agreed offline to handle this pervasive issue in a separate PR.

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+        arg_context="${contexts[0]}";;
+    (*)
+        for context in "${contexts[@]}" ; do # for all contexts
+            if [[ "$context" =~ "$arg_cp" ]] ; then
```
This is substring matching. It is treacherous, in a variety of ways.

francostellari replied:
@MikeSpreitzer you need to explain more, I don't understand the issue.

MikeSpreitzer replied:
Suppose I have contexts named "home" and "homestuck". A substring search for "home" will also match "homestuck".

A little more specifically, it is not uncommon to name something "foo" and later realize that you want another similar thing and name it "foo1" or "foo2".

MikeSpreitzer replied:
"big" is a substring of "bigger"; "near" is a substring of "nearer"; and so on.
You have seen people use names like "big", "bigger", and "biggest".

francostellari replied:
The strict was changed to detect the situation when more than context was being matched, inform the use that the match is not unique, and instruct the suer to explicitly use the `--source-context` argument. Please see the new logic at ~L171:

```bash
                if [[ -n "$arg_source_context" ]] ; then
                    echoerr "there are multiple matching contexts in the source kubeconfig file, specify one with \"--source-context\"!"
                    display_help
                    exit 3
                fi
```

I think this is a good compromise between convenience and clarity

MikeSpreitzer replied:
OK.

francostellari replied:
I added a note in the display_help() function to document the behavior.

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+    exit 1
+fi
+if [[ -z "$arg_context" ]] ; then
+    contexts=($(kubectl --kubeconfig "$arg_kubeconfig" config get-contexts --no-headers -o name 2> /dev/null))
```
I appreciate the work to be convenient, but:
(1) it is undocumented, right? Nobody will rely on it if it is undocumented;
(2) it is unusual --- the usual default is the current context.

I think that it would be fine to default to the current context.

francostellari replied:
I actually designed this way to save of arguments. One can specify the CP alone. Defaulting to the current context is not a good idea in this case because it will be the context of KubeFlex core-hub where you are creating the secret.

What about renaming the argument `--context` to `--adopted-context` to match kflex cli?

To clarify the `--kubeconfig` is also referring to where the adopted context is located. Perhaps that also need to change to be renamed `--adopted-kubeconfig`?

Bottom time currently `--kubeconfig` and `--context` are used to extract/ the secret data.... but not to create the secret object which is using the host kubeconfig/context... currently.

MikeSpreitzer replied:
Maybe something like "source-context" and "source-kubeconfig"? Or "bootstrap-context" and "bootstrap-kubeconfig"? The latter is not great, because they are not where the "bootstrap Secret" go.

francostellari replied:
I renamed the current `--kubeconfig` ==> `--source-kubeconfig` and `--context` ==> `--source-context`

francostellari replied:
I also reworked `--kuneconfig` and `--context` to be the cluster where the secret is creted

MikeSpreitzer replied:
+1

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+        then { arg_kubeconfig="$2"; shift; }
+        else { echo "$0: missing kubeconfig filename" >&2; exit 1; }
+        fi;;
+    (--context|-C)
```
This script is working on two contexts, right? The one that it reads a kubeconfig from, and one that it writes a Secret into. Should both be parameters?

francostellari replied:
I guess so, see above comment.

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+# Create secret
+###############################################################################
+create_ns=true
+for ns in $(kubectl get ns --no-headers -o name) ; do
+    if [[ "namespace/${arg_ns}" == "$ns" ]] ; then
```
This seems like an overly complex way to test whether the namespace exists. Why not just `kubectl get ns "$arg_ns" &> /dev/null` ?

francostellari replied:
yes, sorry for the poor coding

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,195 @@
+        break
+    fi
+done
+if [[ $create_ns == true ]] ; then
```
The `create_ns` variable is read just once. Why not simplify to the following?

```shell
if ! kubectl get ns "$arg_ns" &> /dev/null; then
```

francostellari replied:
yes, see above

MikeSpreitzer reviewed (commented):
I have finished a review. I left independent comments.

francostellari commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +111,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
+
+When the ITS `type` is `external`, the `bootstrapSecret` sub-section can be used to indicate the bootstrap secret used by KubeFlex to connect to the external cluster. Specifically, it can be used to indicate any of the optional values for the name of the secret, the namespace containing the secret, and the name of the key containg the kubeconfig of the external cluster.
+
+If the secret was created using the [create-external-bootstrap-secret.sh](../../../scripts/create-external-bootstrap-secret.sh) script and the value passed to the argument `--controlplane` matches the name of the Control Plane specified by the Helm chart, then the sub-section `bootstrapSecret` is not required because all default values will identify the bootstrap secret created by the script. Mor ...
```
made it absolute to the release

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +115,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
+
+When the ITS `type` is `external`, the `bootstrapSecret` sub-section can be used to indicate the bootstrap secret used by KubeFlex to connect to the external cluster. Specifically, it can be used to change any combination of (a) the name of the secret, (b) the namespace containing the secret, and (c) the name of the key containg the kubeconfig of the external cluster from its default value.
+
+If the secret was created using the [create-external-bootstrap-secret.sh](https://github.com/kubestellar/kubestellar/tree/v$KUBESTELLAR_VERSION/scripts/create-external-bootstrap-secret.sh) script and the value passed to the argument `--controlplane` matches the name of the Control Plane specified by the Helm chart, then the sub-section `bootstrapSecret` is not required because all default values  ...
```
When rendered on the website, that URL will still contain `$KUBESTELLAR_VERSION` (please try it!).
See https://docs.kubestellar.io/release-0.26.0/contribution-guidelines/operations/document-management/#global-variables and look for examples in other source documents.

francostellari replied:
I tried to change it.

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -163,6 +186,36 @@ helm upgrade --install ks-core oci://ghcr.io/kubestellar/kubestellar/core-chart
+
+Note that the last command above creates a secret named `its1-bootstrap` in the `default` namespace of the `kind-kubeflex` cluster.
+
+The `--address` URL needs to be one that the KubeFlex controller can use to open a connection to the external cluster's Kubernetes apiserver(s). In this example, the external cluster is a kind cluster with one kube-apiserver and it listens on port 6443. This example relies on the DNS resolver in Docker networking to map the domain name `ext1-control-plane` to the Docker network address of the con ...
```
Oops, I forgot one critical detail. That kube-apiserver listens on port 6443 _in its node's network namespace_.

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +115,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
-where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be either vcluster (default) or host, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control plane.
+where `name` must specify a name unique among all the control planes in that KubeFlex deployment, the optional `type` can be vcluster (default), host, or external, see [here](https://github.com/kubestellar/kubeflex/blob/main/docs/users.md) for more information, and the optional `install_clusteradm`can be either true (default) or false to enable or disable the installation of OCM in the control pl ...
+
+When the ITS `type` is `external`, the `bootstrapSecret` sub-section can be used to indicate the bootstrap secret used by KubeFlex to connect to the external cluster. Specifically, it can be used to change any combination of (a) the name of the secret, (b) the namespace containing the secret, and (c) the name of the key containg the kubeconfig of the external cluster from their default value.
```
I would say "specify" instead of "change", because change implies that something already exists but we are talking here about specifying the Secret and key (not changing the Secret or key).

francostellari replied:
ok

MikeSpreitzer commented on scripts/create-external-bootstrap-secret.sh:
```diff
@@ -0,0 +1,212 @@
+
+--controlplane|-c <name>       control plane name used to name the secret: <name>-bootstrap
+--namespace|-n <name>          namespace name where to create the secret, default is "default"
+--kubeconfig|-K <filename>     use the specified kubeconfig used to generate the bootstrat secret
```
"generate" is a bit odd here, it calls to mind computation/processing. I think "store" is more to the point.

MikeSpreitzer commented on docs/content/direct/core-chart.md:
```diff
@@ -111,15 +115,27 @@ The third section of the `values.yaml` file allows one to create a list of Inven
     install_clusteradm: true|false  # optional flag to enable/disable the installation of OCM in the control plane (default to true, if not specified)
+    bootstrapSecret: # this section is ignored unless type is "external"
+      name: <secret-name> # default: "<control-plane-name>-bootstrap"
+      namespace: <secret-namespace> # default: "default"
```
Wouldn't it make more sense for the default to be the namespace that the chart is being instantiated in?

francostellari replied:
No for 2 reasons:
1. we do not instruct to install the `core-helm` chart to anything but default
2. this "default" value needs to match the "default" value of the script more than where the chart is installed
I recommend leaving it as it is.

MikeSpreitzer replied:
1. Neither do we forbid instantiating the core Helm chart in another namespace.
2. In my opinion it is better to follow the usual Kubernetes conventions. Two come to mind. One is that for a command, the default namespace to use is the one set in the current context (remember that a context can have a namespace set), if any, otherwise "default". The other is that references from one object to another normally stay in the same namespace. These two play nice together. For example, when using `helm install` without specifying a namespace, the chart will be installed in the namespace determined according to the first rule, and a reference from that chart to a bootstrap secret will reference a Secret in that same namespace when the user does not explicitly specify a namespace.

MikeSpreitzer replied:
FYI:

```console
mspreitz@mjs13 ~ % kubectl config view -o json | jq '.["current-context"] as $cc | .contexts[] | select(.name == $cc) | .namespace // "default"'
"default"
```

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -140,13 +156,20 @@ where `name` must specify a name unique among all the control planes in that Kub
+The local copy of the core chart can be installed in an existing cluster using the commands:
+
+```shell
+helm dependency update core-chart
````
When testing, I got complaint, and it was not clear to me whether this is a fatal problem or not. I checked the exit code, and it claimed success.

```console
mspreitz@mjs13 kubestellar % helm dependency update core-chart
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "minio" chart repository
...Successfully got an update from the "spiffe" chart repository
...Successfully got an update from the "kubestellar" chart repository
...Successfully got an update from the "prometheus-community" chart repository
...Successfully got an update from the "grafana" chart repository
...Successfully got an update from the "bitnami" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Downloading kubeflex-operator from repo oci://ghcr.io/kubestellar/kubeflex/chart
Pulled: ghcr.io/kubestellar/kubeflex/chart/kubeflex-operator:v0.8.0
Digest: sha256:6e889586cbc08d580c100139fce1cf3800f8aed3ce44bec5bee652941d7f5598
Deleting outdated charts
Could not verify core-chart/charts/.DS_Store for deletion: file 'core-chart/charts/.DS_Store' does not appear to be a gzipped archive; got 'application/octet-stream' (Skipping)

mspreitz@mjs13 kubestellar % echo $?
0
```

francostellari replied:
I cannot reproduce your error and I have never seen it before.
It seems to have pulled the latest kubeflex, so I assume it worked.
@pdettori also mentioned that he he needed to do a dependency build to get it to work.
It may be a Helm version issue or a Mac issue.

MikeSpreitzer replied:
Looks like a Mac issue --- `.DS_Store` is a special metadata thing that Darwin adds to a directory when you view it in the GUI.

This complaint was not introduced by this PR.

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -163,6 +186,36 @@ helm upgrade --install ks-core oci://ghcr.io/kubestellar/kubestellar/core-chart
+```shell
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-kind-cluster-with-SSL-passthrough.sh) --name kubeflex --port 9443
+
+bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v$KUBESTELLAR_VERSION/scripts/create-external-bootstrap-secret.sh) --controlplane its1 --context kind-ext1 --address https://ext1-control-plane:6443 --verbose
````
The flag names here need to be updated, right?

MikeSpreitzer commented on docs/content/direct/core-chart.md:
````diff
@@ -140,13 +156,20 @@ where `name` must specify a name unique among all the control planes in that Kub
+
+```shell
+helm dependency update core-chart
+helm upgrade --install core-chart
````
Something is missing on this command line.

```console
mspreitz@mjs13 kubestellar % helm upgrade --install core-chart
Error: "helm upgrade" requires 2 arguments

Usage:  helm upgrade [RELEASE] [CHART] [flags]
```

francostellari replied:
Yes, I forgot the name

MikeSpreitzer reviewed (commented):
I have finished another round of review.
I tested, got as far as instantiating the chart, with only reported issues.

MikeSpreitzer reviewed (commented):
I have completed another round of review.

MikeSpreitzer commented on scripts/kubestellar-snapshot.sh:
I see a lot of duplication of code in this script. There ought to be a way to DRY it out. Not a blocker.

MikeSpreitzer commented:
Testing shows that the test at https://github.com/kubestellar/kubestellar/blob/1ca512d3d99b0fcb3d9eec89b336e5392f8b078d/core-chart/templates/NOTES.txt#L15 also needs to be updated to also exclude "external" type ControlPlanes.

MikeSpreitzer reviewed (approved):
This is good enough to commit as progress. There are a few things to follow up on in following PR(s).

/lgtm
/approve
Thanks!
